nilearn/maskers/nifti_labels_masker.py

```diff
@@ -705,19 +702,76 @@ def transform_single_imgs(self, imgs, confounds=None, sample_mask=None):
             # ids does not include background label
             region_ids[i] = ids[i]
 
+        self.region_names_ = None
         if self.labels is not None:
+
+            # Keep track if background was explicitly passed as a label
+            # TODO:
+            # background should always be explicitly passed in the labels
+            # to avoid this.
+            lower_case_labels = {x.lower() for x in self.labels}
+            knwon_backgrounds = {"background"}
+            background_in_labels = any(
+                knwon_backgrounds.intersection(lower_case_labels)
+            )
+            offset = 1 if background_in_labels else 0
+
+            if len(self.labels) > len(self.labels_) + offset:
+                msg = (
+                    f"\nToo many labels ({len(self.labels)}) passed "
+                    f"for the number of regions ({len(self.labels_)}) "
+                    "in the label image.\n"
+                    f"We should be expect {len(self.labels_) + offset} labels."
+                )
+                if self.resampling_target == "data":
+                    msg += (
+                        "\nNote that this may be due to some regions "
+                        "being dropped from the label image "
+                        "after resampling."
+                    )
+                    warnings.warn(msg, UserWarning, stacklevel=2)
```

Comment:
This warning comes AFTER resampling which is nice to let the users know some regions were dropped, however I think we should have a MUCH earlier check that ensures that the label image has the same number of regions as the number of labels passed.

This kind of check should be done before the labels are resampled otherwise we cannot know if the mismatch between number of labels and regions is due to resampling or if it existed before.